usr/opt/OSCX200/src/man/XI/XGrabDev.man
.\"
.\" *****************************************************************
.\" *                                                               *
.\" *    Copyright (c) Digital Equipment Corporation, 1991, 1994    *
.\" *                                                               *
.\" *   All Rights Reserved.  Unpublished rights  reserved  under   *
.\" *   the copyright laws of the United States.                    *
.\" *                                                               *
.\" *   The software contained on this media  is  proprietary  to   *
.\" *   and  embodies  the  confidential  technology  of  Digital   *
.\" *   Equipment Corporation.  Possession, use,  duplication  or   *
.\" *   dissemination of the software and media is authorized only  *
.\" *   pursuant to a valid written license from Digital Equipment  *
.\" *   Corporation.                                                *
.\" *                                                               *
.\" *   RESTRICTED RIGHTS LEGEND   Use, duplication, or disclosure  *
.\" *   by the U.S. Government is subject to restrictions  as  set  *
.\" *   forth in Subparagraph (c)(1)(ii)  of  DFARS  252.227-7013,  *
.\" *   or  in  FAR 52.227-19, as applicable.                       *
.\" *                                                               *
.\" *****************************************************************
.\"
.\"
.\" HISTORY
.\"
.ds xL Programming With Xlib
.TH XGrabDevice 3X11 "Release 5" "X Version 11" "X FUNCTIONS"
.SH NAME
XGrabDevice, XUngrabDevice \- grab/release the specified extension device
.SH SYNTAX
.nf
\fB
int XGrabDevice\^(\^\fIdisplay\fP, \fIdevice\fP\^, \fIgrab_window\fP\^, \fIowner_events\fP\^, \fIevent_count\fP\^, \fIevent_list\fP\^,
\fIthis_device_mode\fP\^, \fIother_devices_mode\fP\^, \fItime\fP\^)
      Display *\fIdisplay\fP\^;
      XDevice *\fIdevice\fP\^;
      Window \fIgrab_window\fP\^;
      Bool \fIowner_events\fP\^;
      int \fIevent_count\fP\^;
      XEventClass *\fIevent_list\fP\^;
      int \fIthis_device_mode\fP\^, \fIother_devices_mode\fP\^;
      Time \fItime\fP\^;

XUngrabDevice\^(\^\fIdisplay\fP, \fIdevice\fP\^, \fItime\fP\^)
      Display *\fIdisplay\fP\^;
      XDevice *\fIdevice\fP\^;
      Time \fItime\fP\^;
.fi
\fP
.SH ARGUMENTS
.TP 12
.I display
Specifies the connection to the X server.
.TP 12
.I device
Specifies the device to be grabbed or released.
.TP 12
.I grab_window
Specifies the id of a window to be associated with the device.
.TP 12
.I owner_events
Specifies a Boolean value that indicates whether the 
events from the device 
are to be reported as usual or reported with respect to the grab window 
if selected by the event list.
.TP 12
.I event_count
Specifies the number of elements in the event_list array.
.TP 12
.I event_list
Specifies a pointer to a list of event classes that indicates which events
the client wishes to receive.  These event classes must have been
obtained specifying the device being grabbed.
.TP 12
.I this_device_mode
Specifies further processing of events from this device.
You can pass 
\fIGrabModeSync\fP
or
\fIGrabModeAsync\fP.
.TP 12
.I other_devices_mode
Specifies further processing of events from other devices.
You can pass 
\fIGrabModeSync\fP 
or
\fIGrabModeAsync\fP.
.TP 12
.I time
Specifies the time.
You can pass either a timestamp or
\fICurrentTime\fP.
.SH DESCRIPTION
The
\fIXGrabDevice\fP
request actively grabs control of the device and generates
\fIDeviceFocusIn\fP
and
\fIDeviceFocusOut\fP
events.
Further device events are reported only to the
grabbing client.
\fIXGrabDevice\fP
overrides any active device grab by this client.
event_list is a pointer to a list of event classes.  This list indicates
which events the client wishes to receive while the grab is active.
If owner_events is 
\fIFalse\fP , 
all generated device events are reported with
respect to grab_window if selected.
If owner_events is 
\fITrue\fP  
and if a generated
device event would normally be reported to this client, it is reported
normally; otherwise, the event is reported with respect to the
grab_window, and is only reported if specified in the event_list.
.P
If the this_device_mode argument is 
\fIGrabModeAsync\fP ,
device event processing continues
as usual. 
If the device is currently frozen by this client, 
then processing of device events is resumed.
If the this_device_mode  argument is
\fIGrabModeSync\fP ,
the state of the device (as seen by client applications) appears to freeze,
and the X server generates no further device events until the
grabbing client issues a releasing 
\fIXAllowDeviceEvents\fP 
call or until the device grab is released.
Actual device changes are not lost while the device is frozen; 
they are simply queued in the server for later processing.
.P
If other_devices_mode is 
\fIGrabModeAsync\fP ,
processing of events from other devices is unaffected
by activation of the grab.  
If other_devices_mode is 
\fIGrabModeSync\fP,
the state of all devices except the grabbed device
 (as seen by client applications) appears to freeze, 
and the X server generates no further events from those devices
until the grabbing client issues a releasing 
\fIXAllowDeviceEvents\fP 
call or until the device grab is released.
Actual events are not lost while the devices are frozen; 
they are simply queued in the server for later processing.
.P
If the device is actively grabbed by some other client,
\fIXGrabDevice\fP
fails and returns
\fIAlreadyGrabbed\fP.
If grab_window is not viewable,
it fails and returns
\fIGrabNotViewable\fP.
If the device is frozen by an active grab of another client,
it fails and returns
\fIGrabFrozen\fP.
If the specified time is earlier than the last-device-grab time 
or later than the current X server time,
it fails and returns
\fIGrabInvalidTime\fP.
Otherwise, the last-device-grab time is set to the specified time
\fI( CurrentTime\fP 
is replaced by the current X server time).
.LP
If a grabbed device is closed by a client while an active grab by that
client is in effect, the active grab is released.  If the device is
frozen only by an active grab of the requesting client, it is thawed.
.P
\fIXGrabDevice\fP
can generate
\fIBadClass\fP, \fIBadDevice\fP, \fIBadValue\fP, and \fIBadWindow\fP 
errors.
.P
The
\fIXUngrabDevice\fP
request
releases the device and any queued events if this client has it actively 
grabbed from either
\fIXGrabDevice\fP
or
\fIXGrabDeviceKey\fP.
If other devices are frozen by the grab, \fIXUngrabDevice\fP thaws them.
\fIXUngrabDevice\fP
does not release the device and any queued events
if the specified time is earlier than
the last-device-grab time or is later than the current X server time.
It also generates
\fIDeviceFocusIn\fP 
and 
\fIDeviceFocusOut\fP 
events.
The X server automatically performs an 
\fIUngrabDevice\fP 
request if the event window for an
active device grab becomes not viewable.
.LP
\fIXUngrabDevice\fP
can generate a \fIBadDevice\fP error.
.SH DIAGNOSTICS
.TP 12
\fIBadDevice\fP
An invalid device was specified.  The specified device does not exist or has 
not been opened by this client via \fIXOpenInputDevice\fP.  This error may
also occur if the specified device is the X keyboard or X pointer device.
.TP 12
\fIBadValue\fP
Some numeric value falls outside the range of values accepted by the request.
Unless a specific range is specified for an argument, the full range defined
by the argument's type is accepted.  Any argument defined as a set of
alternatives can generate this error.
.TP 12
\fIBadWindow\fP
A value for a Window argument does not name a defined Window.
.SH "SEE ALSO"
XAllowDeviceEvents(3X),
XGrabDeviceButton(3X),
XGrabDeviceKey(3X),
.br
\fI\*(xL\fP


usr/opt/OSCX200/src/extensions/server/xinput/xgrabdev.c
/*
 * *****************************************************************
 * *                                                               *
 * *    Copyright (c) Digital Equipment Corporation, 1991, 1994    *
 * *                                                               *
 * *   All Rights Reserved.  Unpublished rights  reserved  under   *
 * *   the copyright laws of the United States.                    *
 * *                                                               *
 * *   The software contained on this media  is  proprietary  to   *
 * *   and  embodies  the  confidential  technology  of  Digital   *
 * *   Equipment Corporation.  Possession, use,  duplication  or   *
 * *   dissemination of the software and media is authorized only  *
 * *   pursuant to a valid written license from Digital Equipment  *
 * *   Corporation.                                                *
 * *                                                               *
 * *   RESTRICTED RIGHTS LEGEND   Use, duplication, or disclosure  *
 * *   by the U.S. Government is subject to restrictions  as  set  *
 * *   forth in Subparagraph (c)(1)(ii)  of  DFARS  252.227-7013,  *
 * *   or  in  FAR 52.227-19, as applicable.                       *
 * *                                                               *
 * *****************************************************************
 */
/*
 * HISTORY
 */
/* $XConsortium: xgrabdev.c,v 1.10 90/05/18 15:32:29 rws Exp $ */

/************************************************************
Copyright (c) 1989 by Hewlett-Packard Company, Palo Alto, California, and the 
Massachusetts Institute of Technology, Cambridge, Massachusetts.

			All Rights Reserved

Permission to use, copy, modify, and distribute this software and its
documentation for any purpose and without fee is hereby granted,
provided that the above copyright notice appear in all copies and that
both that copyright notice and this permission notice appear in
supporting documentation, and that the names of Hewlett-Packard or MIT not be
used in advertising or publicity pertaining to distribution of the
software without specific, written prior permission.

HEWLETT-PACKARD DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING
ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO EVENT SHALL
HEWLETT-PACKARD BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR
ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION,
ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
SOFTWARE.

********************************************************/

/***********************************************************************
 *
 * Extension function to grab an extension device.
 *
 */


#define	 NEED_EVENTS
#define	 NEED_REPLIES
#include "X.h"				/* for inputstr.h    */
#include "Xproto.h"			/* Request macro     */
#include "inputstr.h"			/* DeviceIntPtr	     */
#include "windowstr.h"			/* window structure  */
#include "XI.h"
#include "XIproto.h"

extern	int 		IReqCode;
extern	int		BadDevice;
extern	int		BadClass;
extern	InputInfo	inputInfo;
extern	XExtEventInfo	EventInfo[];
extern	void		(* ReplySwapVector[256]) ();
DeviceIntPtr		LookupDeviceIntRec();

/***********************************************************************
 *
 * Swap the request if the requestor has a different byte order than us.
 *
 */

int
SProcXGrabDevice(client)
    register ClientPtr client;
    {
    register char n;
    register long *p;
    register int i;

    REQUEST(xGrabDeviceReq);
    swaps(&stuff->length, n);
    swapl(&stuff->grabWindow, n);
    swapl(&stuff->time, n);
    swaps(&stuff->event_count, n);
    p = (long *) &stuff[1];
    for (i=0; i<stuff->event_count; i++)
        {
        swapl(p, n);
	p++;
        }

    return(ProcXGrabDevice(client));
    }

/***********************************************************************
 *
 * Grab an extension device.
 *
 */

int
ProcXGrabDevice(client)
    ClientPtr client;
    {
    int			error;
    xGrabDeviceReply 	rep;
    DeviceIntPtr 	dev;
    struct tmask	tmp[EMASKSIZE];

    REQUEST(xGrabDeviceReq);
    REQUEST_AT_LEAST_SIZE(xGrabDeviceReq);

    if (stuff->length !=(sizeof(xGrabDeviceReq)>>2) + stuff->event_count)
	{
	SendErrorToClient (client, IReqCode, X_GrabDevice, 0, BadLength);
	return Success;
	}

    rep.repType = X_Reply;
    rep.RepType = X_GrabDevice;
    rep.sequenceNumber = client->sequence;
    rep.length = 0;

    dev = LookupDeviceIntRec (stuff->deviceid);
    if (dev == NULL)
	{
	SendErrorToClient(client, IReqCode, X_GrabDevice, 0, BadDevice);
	return Success;
	}

    if (CreateMaskFromList (client, (XEventClass *)&stuff[1], 
	stuff->event_count, tmp, dev, X_GrabDevice) != Success)
	return Success;

    error = GrabDevice (client, dev, stuff->this_device_mode, 
	stuff->other_devices_mode, stuff->grabWindow, stuff->ownerEvents, 
	stuff->time, tmp[stuff->deviceid].mask, &rep.status);

    if (error != Success)
	{
	SendErrorToClient(client, IReqCode, X_GrabDevice, 0, error);
	return Success;
	}
    WriteReplyToClient(client, sizeof(xGrabDeviceReply), &rep);
    return Success;
    }


/***********************************************************************
 *
 * This procedure creates an event mask from a list of XEventClasses.
 *
 */

int
CreateMaskFromList (client, list, count, mask, dev, req)
    ClientPtr		client;
    XEventClass		*list;
    int			count;
    struct tmask	mask[];
    DeviceIntPtr	dev;
    int			req;
    {
    int			i,j;
    int			device;
    DeviceIntPtr	tdev;
    extern int		ExtEventIndex;

    for (i=0; i<EMASKSIZE; i++)
	{
	mask[i].mask = 0;
	mask[i].dev = NULL;
	}

    for (i=0; i<count; i++, list++)
	{
	device = *list >> 8;
	tdev = LookupDeviceIntRec (device);
	if (tdev==NULL || (dev != NULL && tdev != dev))
	    {
	    SendErrorToClient(client, IReqCode, req, 0, BadClass);
	    return BadClass;
	    }

	for (j=0; j<ExtEventIndex; j++)
	    if (EventInfo[j].type == (*list & 0xff))
		{
		mask[device].mask |= EventInfo[j].mask;
		mask[device].dev = (Pointer) tdev;
		break;
		}
	}
    return Success;
    }

/***********************************************************************
 *
 * This procedure writes the reply for the XGrabDevice function,
 * if the client and server have a different byte ordering.
 *
 */

SRepXGrabDevice (client, size, rep)
    ClientPtr	client;
    int		size;
    xGrabDeviceReply	*rep;
    {
    register char n;

    swaps(&rep->sequenceNumber, n);
    swapl(&rep->length, n);
    WriteToClient(client, size, rep);
    }
